### PR TITLE
Enhancing text to speech

### LINKED ISSUE #108: Review and test text-to-speech
### Description

This issue is to track feedback for the senselab text-to-speech task.

### Tasks

- [x] Review tutorial
- [x] Review documentation
- [x] Review code
- [x] Test edge cases

### Freeform Notes
Fabio's documentation guidelines:
- Documentation - Considering the tasks that you have implemented, work on the documentation
- Explain what the task is about, give an overview of the metrics generally used to evaluate the task, report some existing benchmarks in the literature, if any
- Make docstrings in Google style and consistent with with the CONTRIBUTING.md document

Notebook tutorial

- Prepare a jupyter notebook tutorial for each of your tasks. This should show how to use the senselab interface that you have implemented, both with and without using pydra

### PR DESCRIPTION
**Pull request by fabiocat93**
- added mars5tts
- added tutorial
- improved documentation
- addressed @900miles's review

### CODE REVIEW

**Comment by fabiocat93** [comment]
> There are no tutorials for this task. There is no explanation for what the task is about, what the metrics are for evaluating the tasks, nor any benchmarks in the literature. There are no examples in the docstring of how the function should be used. The doctring says that model should be of type SenselabModel but the code requires it to be of type HFModel.

@900miles documentation was added and docstrings adjusted

**Comment by fabiocat93** [comment]
> When running synthesize_texts, I get the following warning:  
"The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:10000 for open-end generation." 
This seems to happen once per text synthesized.


@900miles this is a warning that is specific to the bark model. It is nothing we should be worried about for now

**Comment by fabiocat93** [comment]
> The supported models seems to be incorrect. The documentation is vague, but seems to imply that any huggingface tts model is supported. However, I have tried microsoft/speecht5_tts, microsoft/speecht5_tts, fishaudio/fish-speech-1.2-sft, and facebook/mms-tts-eng and all have failed with various errors. suno/bark is the only one I have tried so far that works.


@900miles I have now specified some of the models that we recommend. I have also clarified that not every model is supported, despite they wrongly labeled it with the text-to-speech label. microsoft/speecht5_tts, facebook/mms-tts-eng, suno/bark are in the tests/tutorial and work fine. fishaudio/fish-speech-1.2-sft was never able to make it work

**Comment by fabiocat93** [comment]
> The tests seem to be missing two cases: 1. testing with different models other than the default and 2. testing the case of synthesizing more than one text at one time.

@900miles fixed

**Comment by fabiocat93** [comment]
@900miles this should be ready for a second review whenever you have time

**Comment by fabiocat93** [comment]
@wilke0818 any progress update with styletts2?

File: src/senselab/audio/tasks/text_to_speech/api.py
Comment (900miles) [inline src/senselab/audio/tasks/text_to_speech/api.py]: A bit confusing but it looks like the code also accepts a list of Audios, not just a list of tuples.

File: src/senselab/audio/tasks/text_to_speech/api.py
Comment (900miles) [inline src/senselab/audio/tasks/text_to_speech/api.py]: The documentation says model needs to be of type `HFModel` but the code allows it to be of type `HFModel` OR `TorchModel` and the function header allows any `SenselabModel`. These should be reconciled.

File: src/senselab/audio/tasks/text_to_speech/api.py
Comment (900miles) [reply src/senselab/audio/tasks/text_to_speech/api.py]: The code also allows target to be None. A couple words on what's going on might be helpful. Maybe something saying that how you specify target depends on which model is being used.

File: src/senselab/audio/tasks/text_to_speech/api.py
Comment (900miles) [inline src/senselab/audio/tasks/text_to_speech/api.py]: "The lists of texts and target audios must have the same length." or something along those lines

File: src/senselab/audio/tasks/text_to_speech/api.py
Comment (900miles) [inline src/senselab/audio/tasks/text_to_speech/api.py]: "The lists of texts and target transcripts must have the same length." or something along those lines

File: src/senselab/audio/tasks/text_to_speech/api.py
Comment (900miles) [inline src/senselab/audio/tasks/text_to_speech/api.py]: Why do you unzip `target` and then rezip `target` instead of just using `target` directly here?

File: src/senselab/audio/tasks/text_to_speech/doc.md
Comment (900miles) [inline src/senselab/audio/tasks/text_to_speech/doc.md]: It's a little unclear what "[it] provides text-to-speech capabilities with a target voice (and its transcript) in English" means exactly.

File: src/senselab/audio/tasks/text_to_speech/marstts.py
Comment (900miles) [inline src/senselab/audio/tasks/text_to_speech/marstts.py:41]: This requires `language="en"`, which will make the function throw an error with the default setting (`language=None`)

File: src/senselab/audio/tasks/text_to_speech/marstts.py
Comment (900miles) [inline src/senselab/audio/tasks/text_to_speech/marstts.py:130]: These errors should specify which audio is the offending audio.

**Review by 900miles — COMMENTED** [review]
Looks good so far, I've left some comments. I've still got a few files and the tutorial to review.

File: src/senselab/audio/tasks/text_to_speech/api.py
Comment (fabiocat93) [reply src/senselab/audio/tasks/text_to_speech/api.py]: fair. i unzip it to make some checks that are not specific to the model. and then, if pass, I structure the target in the format accepted by the mars5tts function. i see how this is not really optimized, any suggestion is welcome

**Comment by fabiocat93** [comment]
thank you @900miles , i think i have addressed your comments. 
@wilke0818 after yesterday's meeting, what do we want to do with `styletts2`? i honestly think that, if we go with a docker-based solution, we will create some challenge for entry-level people who want to start learning about audio signal processing and this is something we don't want of course

**Comment by wilke0818** [comment]
@fabiocat93 I am not sure how hard it would be for new users as theoretically the Docker would be under the hood. However, as to not delay these other tools and in order to give us more time to think about this, I would recommend reverting my commit, I don't know if it is easily revertible since you've committed more, but we could honestly remove the StyleTTS code (though perhaps make sure to push it into a seperate branch so it doesn't have to be rewritten). Do you want to do this? Or I can but it might be near the end of the day.

**Comment by fabiocat93** [comment]
> @fabiocat93 I am not sure how hard it would be for new users as theoretically the Docker would be under the hood. However, as to not delay these other tools and in order to give us more time to think about this, I would recommend reverting my commit, I don't know if it is easily revertible since you've committed more, but we could honestly remove the StyleTTS code (though perhaps make sure to push it into a seperate branch so it doesn't have to be rewritten). Do you want to do this? Or I can but it might be near the end of the day.

That makes sense. I don't have a buffer for doing it today. I would appreciate it if you could do it. Otherwise, I can find some time tomorrow night.
That said, let's make sure we continue the docker-related discussion in the new PR thread

File: pyproject.toml
Comment (wilke0818) [inline pyproject.toml:59]: was this for Mars or StyleTTS?

File: src/senselab/audio/tasks/forced_alignment/forced_alignment.py
Comment (wilke0818) [inline src/senselab/audio/tasks/forced_alignment/forced_alignment.py:306]: why is this in this PR?

File: src/senselab/audio/tasks/speaker_embeddings/speechbrain.py
Comment (wilke0818) [inline src/senselab/audio/tasks/speaker_embeddings/speechbrain.py:95]: why is this in this PR?

File: src/senselab/audio/tasks/text_to_speech/__init__.py
Comment (wilke0818) [inline src/senselab/audio/tasks/text_to_speech/__init__.py:1]: If it becomes common for us to use the doc.md method then let's add D415 to the ignore list.

File: src/senselab/audio/tasks/text_to_speech/huggingface.py
Comment (wilke0818) [inline src/senselab/audio/tasks/text_to_speech/huggingface.py:52]: how do these differ from simple **kwargs?

File: src/senselab/audio/tasks/text_to_speech/styletts2.py
Comment (wilke0818) [inline src/senselab/audio/tasks/text_to_speech/styletts2.py:1]: this file is generally unused but I think keeping it as an additional example of how to go about integrating new models is good. Also eventually it would be nice to get this integrated if possible.

File: src/senselab/utils/tasks/cross_correlation.py
Comment (wilke0818) [inline src/senselab/utils/tasks/cross_correlation.py:34]: again, is this supposed to be in this PR or is it just a weird Github thing happening?

File: src/tests/audio/tasks/text_to_speech_test.py
Comment (wilke0818) [inline src/tests/audio/tasks/text_to_speech_test.py:71]: this test takes a really long time but not sure if there is anything we can do about that.

**Review by wilke0818 — APPROVED** [review]
Assuming @900miles reviews have been addressed, I think this is good to go

File: pyproject.toml
Comment (fabiocat93) [reply pyproject.toml:59]: mars5

File: src/senselab/audio/tasks/forced_alignment/forced_alignment.py
Comment (fabiocat93) [reply src/senselab/audio/tasks/forced_alignment/forced_alignment.py:306]: good question. that's my bad. probably somehow there was a style issue in the main branch and I fixed it since couldn't commit otherwise

File: src/senselab/audio/tasks/speaker_embeddings/speechbrain.py
Comment (fabiocat93) [reply src/senselab/audio/tasks/speaker_embeddings/speechbrain.py:95]: same as above (mypy issue)

**Comment by 900miles** [comment]
Sorry, haven't had a chance to finish my review yet. I'll get it done by end-of-day today

File: src/senselab/audio/tasks/text_to_speech/__init__.py
Comment (fabiocat93) [reply src/senselab/audio/tasks/text_to_speech/__init__.py:1]: I would recommend keeping track of how often this method is used and evaluating if it’s truly beneficial before making the change permanent

File: src/senselab/audio/tasks/text_to_speech/huggingface.py
Comment (fabiocat93) [reply src/senselab/audio/tasks/text_to_speech/huggingface.py:52]: not really different. if you want, i can make it with `**kwargs`

File: src/senselab/audio/tasks/text_to_speech/styletts2.py
Comment (fabiocat93) [reply src/senselab/audio/tasks/text_to_speech/styletts2.py:1]: maybe let's add a comment specifying that this is in progress somewhere?

File: src/senselab/utils/tasks/cross_correlation.py
Comment (fabiocat93) [reply src/senselab/utils/tasks/cross_correlation.py:34]: same as above. sorry

File: src/tests/audio/tasks/text_to_speech_test.py
Comment (fabiocat93) [reply src/tests/audio/tasks/text_to_speech_test.py:71]: i just think that the problem is with it running very inefficiently on the CPU....

**Comment by fabiocat93** [comment]
> Sorry, haven't had a chance to finish my review yet. I'll get it done by end-of-day today

no prob

File: tutorials/text_to_speech.ipynb
Comment (900miles) [inline tutorials/text_to_speech.ipynb]: "which just requires as input ..."

File: tutorials/text_to_speech.ipynb
Comment (900miles) [inline tutorials/text_to_speech.ipynb]: We should add a few words on how the user can extract their own speaker embeddings and link to our docs on speaker embeddings

File: tutorials/text_to_speech.ipynb
Comment (wilke0818) [inline tutorials/text_to_speech.ipynb:171]: targets is supposed to be a tuple, not a list. code errors.

Comment (900miles) [reply tutorials/text_to_speech.ipynb:171]: I get the same error. Line 239

File: tutorials/text_to_speech.ipynb
Comment (900miles) [inline tutorials/text_to_speech.ipynb:238]: This line took 14 minutes to generate, is that expected?

File: src/senselab/audio/tasks/text_to_speech/api.py
Comment (900miles) [reply src/senselab/audio/tasks/text_to_speech/api.py]: Could you just do the checks without creating new variables / functions to unzip and rezip? Like have the loop in the beginning just do the checks, and if they pass, send `target` into `mars5tts`?

**Review by 900miles — CHANGES_REQUESTED** [review]
Looks great just a couple more comments

File: tutorials/text_to_speech.ipynb
Comment (wilke0818) [reply tutorials/text_to_speech.ipynb:238]: i think that's about what I had when running on CPU. don't think there is much that can be done about that.

File: src/senselab/audio/tasks/text_to_speech/api.py
Comment (fabiocat93) [reply src/senselab/audio/tasks/text_to_speech/api.py]: @900miles the problem with the approach you suggest is that you cannot guarantee data type consistency and mypy complains about that:
`Argument "target" to "synthesize_texts_with_mars5tts" of "Mars5TTS" has incompatible type "list[tuple[Audio, str | None]] | None"; expected "list[tuple[Audio, str]]"  [arg-type]`
zipping `target` again is a workaround to make the code work and `mypy` happy

File: tutorials/text_to_speech.ipynb
Comment (fabiocat93) [reply tutorials/text_to_speech.ipynb:238]: fyi, it took 20 seconds on my laptop

File: src/senselab/audio/tasks/text_to_speech/api.py
Comment (wilke0818) [reply src/senselab/audio/tasks/text_to_speech/api.py]: I can take a look at this later. there are definitely approaches that will make mypy happy that don't require zipping/unzipping

File: tutorials/text_to_speech.ipynb
Comment (wilke0818) [reply tutorials/text_to_speech.ipynb:238]: do you have a GPU? On GPU/running a large batch of varied audios, I averaged about 25 seconds per audio. This isn't bad but will be hard to scale with, especially since we haven't been able to show performance improvements by splitting across multiple GPUs

Comment (fabiocat93) [reply tutorials/text_to_speech.ipynb:238]: I have no GPU on my macbook and I request CPU as a preferred device

Comment (wilke0818) [reply tutorials/text_to_speech.ipynb:238]: then to be honest I have no idea how it ran so fast since @900miles and I have both used the code and found it can take up to 15 minutes.

Comment (fabiocat93) [reply tutorials/text_to_speech.ipynb:238]: @wilke0818 sorry, 20 seconds was bark-small. mars5tts takes 8 seconds for the initialization and 4 minutes and 36 seconds for synthesizing "Hello, world. It's nice to meet you." on the CPU

Comment (900miles) [reply tutorials/text_to_speech.ipynb:238]: Hmm.. Perhaps a warning that it might take a while? So users don't think something is broken after waiting 5–15 minutes.

Comment (900miles) [reply tutorials/text_to_speech.ipynb:238]: I'm not sure why yours is so much faster. I ran it on openmind, I don't know what kind of cpus they have

Comment (wilke0818) [reply tutorials/text_to_speech.ipynb:238]: I wouldn't be surprised if newer Apple's (what generation is yours @fabiocat93) have some dedicated GPU(like) hardware that speeds it up

Comment (fabiocat93) [reply tutorials/text_to_speech.ipynb:238]: M2

Comment (wilke0818) [reply tutorials/text_to_speech.ipynb:238]: yeah I would say the M2 simply just is a higher performance chip compared to most Intel chips. It would be interesting to configure actual tests/benchmarks for this and other tasks.

**Comment by wilke0818** [comment]
The TypeGuard should help address the issue that @fabiocat93's zip strategy was working around. Basically MyPy has no real ability to allow for checking Iterables, especially when we allow them to have mixed types, so this is a simple workaround for it. Some other notes:
- I changed `target` to `targets` because it was bothering me
- I made `targets` into `Optional[List[Audio | Tuple[Audio, str]]]` since it seems to me like we have a few different cases, a) no targets provided so Optional, b) targets are all just Audios, c) targets need to be strictly Audios and their transcripts like with Mars5-TTS, and d) if we ever get StyleTTS integrated, the Audio is required with optional transcripts so rather than forcing that choice on the user it seems better to just have internal logic choose which to use
- Technically `targets` needs to be the near last parameter to `synthesize_texts` or else it isn't backwards compatible for users who might've been using purely positional arguments instead of keywords.

**Comment by fabiocat93** [comment]
> The TypeGuard should help address the issue that @fabiocat93's zip strategy was working around. Basically MyPy has no real ability to allow for checking Iterables, especially when we allow them to have mixed types, so this is a simple workaround for it. Some other notes:
> 
> * I changed `target` to `targets` because it was bothering me
> * I made `targets` into `Optional[List[Audio | Tuple[Audio, str]]]` since it seems to me like we have a few different cases, a) no targets provided so Optional, b) targets are all just Audios, c) targets need to be strictly Audios and their transcripts like with Mars5-TTS, and d) if we ever get StyleTTS integrated, the Audio is required with optional transcripts so rather than forcing that choice on the user it seems better to just have internal logic choose which to use
> * Technically `targets` needs to be the near last parameter to `synthesize_texts` or else it isn't backwards compatible for users who might've been using purely positional arguments instead of keywords.

thank you @wilke0818. looks good to me. let's wait for @900miles (final) review

**Review by 900miles — APPROVED** [review]
Looks great! Thanks @fabiocat93 and @wilke0818

**Pull request merged**
